### ISSUE TITLE
Skipped rewrite bug

### ISSUE THREAD

**Issue by alexmojaki** [ssbr/refex#20]
I came across this blog post: https://engineering.soroco.com/abstract-syntax-tree-for-patching-code-and-assessing-code-quality/

I thought this sounded like a perfect use case for refex, so for fun I tried it on the first example:

```python
import pandas as pd

mi = pd.MultiIndex.from_product([[1, 2], ['a', 'b']], names=['x', 'y'])
print(mi.levels[0].name)

mi.levels[0].name = "new name"
print(mi.levels[0].name)
```

```bash
$ refex --mode=py.stmt '$var.levels[$idx].name = $val' --sub '$var = $var.set_names($val, level=$idx)' script.py
Skipped rewrite: Bug in Python formatter? Even "safe" formatting of Python template produced an incorrect and different AST, so it must be discarded:  template='$var = $var.set_names($val, level=$idx)', substitute(matches for {'val': '"new name"', 'idx': '0', 'var': 'mi', '__root': 'mi.levels[0].name = "new name"'}) -> 'var = (mi).set_names(("new name"), level=(0))'
```

Looks like it's confused by `$var = $var...`, it wrote `var = mi...` instead of `mi = mi...`.

**Comment by ssbr** [comment]
`--mode=py.stmt` is definitely less tested than `py.expr`, but assignment really ought to work, and you're right that Refex appears to be choking on the assignment target specifically for some reason. I'll take a look at this on Monday.

(P.S. somebody knows about refex outside of work?!? Woo!)

**Comment by ssbr** [comment]
**tl;dr:** great find, fix incoming!

I lied, ended up doing it today because I was so excited to have a user, and so disappointed it didn't work for that user. :)

The bug is that the lexical template class doesn't understand multiple substitutions of the same variable -- if you substitute x=1 into `($x, $x)`, it becomes `(x, 1)` (due to peculiarities of the internal representation of the lexical template -- it actually saves the 'x' token so that it could replace it with a different token). So it is not specific to `py.stmt`, and in fact I'm a bit surprised it took until now to uncover!

This affects the safety check, because the implementation involves parenthesizing `$var`, which occurs multiple times in the replacement template, and thus falls victim to this failed replacement. It then correctly observes that the substitution doesn't look like the original template, and bails.

I also, in the process of investigating this, discovered a bug that _is_ specific to assignment -- if your variable comes from an lvalue, it doesn't substitute into rvalues. (Not guaranteeing this specific example works, but something like `$x = 1` being replaced by `foo($x)`) That's also going to be fixed.

I'm mailing out the chain of fixes, they will not render nicely inside github because copybara doesn't understand chained commits. (Inside of Google where I did the change, this is a sequence of 3 PRs, each of which builds on the previous one -- github doesn't have that workflow, nor does copybara support it.)

(I guess I should also update PyPI after this, so that makes a chain of 4 commits.)

**Comment by ssbr** [comment]
Tested, and indeed, #23 should fix this bug:

```
$ cat > /tmp/foo.py
import pandas as pd

mi = pd.MultiIndex.from_product([[1, 2], ['a', 'b']], names=['x', 'y'])
print(mi.levels[0].name)

mi.levels[0].name = "new name"
print(mi.levels[0].name)
$ [refex command] --mode=py.stmt '$var.levels[$idx].name = $val' --sub '$var = $var.set_names($val, level=$idx)' -i /tmp/foo.py
/tmp/foo.py
-mi.levels[0].name = "new name"
+mi = mi.set_names("new name", level=0)
```

**Comment by alexmojaki** [comment]
Haha, I remember it from discussions in the asttokens repo. I've never actually used it but I try to remember it and keep an eye out for good opportunities.

If you release this fix, maybe comment about it on https://www.reddit.com/r/Python/comments/nstf0t/i_learnt_to_use_asts_to_patch_100000s_lines_of

**Comment by ssbr** [comment]
Yes, CL chains in Google get represented *extremely* poorly in copybara. Wowee.

OK, I'm going to release this on PyPI (i.e.: reread the poetry docs because I forget) and mark as fixed.

And then post comments on all the HN and reddit threads yes. ;)
